### ISSUE TITLE
GitHub test action sometimes fails on VNF entry deletion

### ISSUE THREAD

**Issue by PlagueCZ** [ironcore-dev/dpservice#513]
On rare instances (currently it seems that something in that the time window must align for this to be at least sometimes reproducible by re-running the `test` action), the tester docker image fails on prefix or lbprefix deletion.

Posting the lbprefix deletion because its code does less actions than the alias prefix counterpart, yet the error is the same:
```
----------------------------- Captured stdout call -----------------------------
dpservice-cli add lbprefix --interface-id=vm2 --prefix=2a10:defe:e01f:f4::2/128
2024-04-04 13:27:15.318 2(grpc) I GRPC: Adding loadbalancer target prefix, interface_id: vm2, prefix: 2a10:defe:e01f:f4::2, prefix_len: 128 [../src/grpc/dp_async_grpc.cpp:857:FillRequest()]
 > {"kind":"LoadBalancerPrefix","metadata":{"interface_id":"vm2"},"spec":{"prefix":"2a10:defe:e01f:f4::2/128","underlay_route":"fc00:1::7f:0:fb"},"status":{"code":0,"message":"Success"}}
dpservice-cli list lbprefixes --interface-id=vm2
2024-04-04 13:27:15.326 2(grpc) I GRPC: Listing loadbalancer target prefixes, interface_id: vm2 [../src/grpc/dp_async_grpc.cpp:899:FillRequest()]
 > {"kind":"LoadBalancerPrefixList","metadata":{"interface_id":"vm2"},"status":{"code":0,"message":"Success"},"items":[{"kind":"LoadBalancerPrefix","metadata":{"interface_id":"vm2"},"spec":{"prefix":"2a10:defe:e01f:f4::2/128","underlay_route":"fc00:1::7f:0:fb"},"status":{"code":0,"message":""}}]}
dpservice-cli add lbprefix --interface-id=invalid_name --prefix=2a10:defe:e01f:f4::2/128
2024-04-04 13:27:15.338 2(grpc) I GRPC: Adding loadbalancer target prefix, interface_id: invalid_name, prefix: 2a10:defe:e01f:f4::2, prefix_len: 128 [../src/grpc/dp_async_grpc.cpp:857:FillRequest()]
2024-04-04 13:27:15.338 3(worker) W GRPC: Failed request, grpc_request: 30, grpc_error: 205, grpc_message: NO_VM [../src/grpc/dp_grpc_impl.c:1146:dp_process_request()]
 > {"kind":"LoadBalancerPrefix","metadata":{"interface_id":"invalid_name"},"spec":{"prefix":"2a10:defe:e01f:f4::2/128"},"status":{"code":205,"message":"NO_VM"}}
dpservice-cli del lbprefix --interface-id=invalid_name --prefix=2a10:defe:e01f:f4::2/128
2024-04-04 13:27:15.350 2(grpc) I GRPC: Removing loadbalancer target prefix, interface_id: invalid_name, prefix: 2a10:defe:e01f:f4::2, prefix_len: 128 [../src/grpc/dp_async_grpc.cpp:880:FillRequest()]
2024-04-04 13:27:15.350 3(worker) W GRPC: Failed request, grpc_request: 31, grpc_error: 205, grpc_message: NO_VM [../src/grpc/dp_grpc_impl.c:1146:dp_process_request()]
 > {"kind":"LoadBalancerPrefix","metadata":{"interface_id":"invalid_name"},"spec":{"prefix":"2a10:defe:e01f:f4::2/128"},"status":{"code":205,"message":"NO_VM"}}
dpservice-cli del lbprefix --interface-id=vm2 --prefix=2a10:defe:e01f:f4::2/128
2024-04-04 13:27:15.358 2(grpc) I GRPC: Removing loadbalancer target prefix, interface_id: vm2, prefix: 2a10:defe:e01f:f4::2, prefix_len: 128 [../src/grpc/dp_async_grpc.cpp:880:FillRequest()]
2024-04-04 13:27:15.358 3(worker) W GRPC: Cannot lookup VNF to delete, ipv6: fb00:1::7f:0:fb, error: -2, errmsg: No such file or directory [../src/dp_vnf.c:177:dp_del_vnf()]
2024-04-04 13:27:15.358 3(worker) W GRPC: Failed request, grpc_request: 31, grpc_error: 201, grpc_message: NOT_FOUND [../src/grpc/dp_grpc_impl.c:1146:dp_process_request()]
 > {"kind":"LoadBalancerPrefix","metadata":{"interface_id":"vm2"},"spec":{"prefix":"2a10:defe:e01f:f4::2/128"},"status":{"code":201,"message":"NOT_FOUND"}}
=========================== short test summary info ============================
FAILED test_zzz_grpc.py::test_grpc_add_list_delLoadBalancerTargets_ipv6 - grp...
!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!
=================== 1 failed, 59 passed, 5 skipped in 26.35s ===================
```

In every instance of this bug, you can see the same thing in the output, i.e. that instead of trying to find `fc00:1::` prefixed VNF address (as configured by command-line parameter to dpservice), it is trying to find `fb00:1::` prefix with the right suffix. 

But this only happens sparingly, which seems to indicate that "something" "sometimes" must switch three bits! Using `vnf_ul_addr6[0] ^= 0x7;` at `dp_vnf.c:223` produces the same error in output, but for this operation to happen on its own and only sometimes is really strange.

We (me and @byteocean ) have added asserts to the code at few places in `fix/strange`.

We were successful at running 7796137db0a98bee0bfef1f6136720bf7370bacd and seeing the error, but the code did not assert, so that is a partial confirmation, that the address is being sent to the hashtable in the right form.

The latest commit unfortunately does not fail. I cannot even reproduce the bug again by re-running the old tests, implying that this is dependant on the hardware state of the runner itself.

But this also happened a month ago once, so it's not a one-time occurrence.

**Comment by PlagueCZ** [comment]
I "successfully" ran a failing test of 1bd8426, results are [here](https://github.com/ironcore-dev/dpservice/actions/runs/8572188910/job/23494145737). Most imporantly there is an assertion in `dp_vnf.c:239` which is `dp_del_vnf_by_value()`, the assertion happes on delet, not while adding.

Given present assertions I added to the code, the address must be changing after it is successfully saved into the hash table and also then looked-up, see https://github.com/ironcore-dev/dpservice/blob/1bd84263d761efb0a7ebed32086531c434896ba0/src/dp_vnf.c#L63

My previous idea of error happening while storing the value is thus proven false.

**Comment by PlagueCZ** [comment]
I added debug outputs to check the actual pointer values in 71d1fef4d673cc59ac4d24724e46355feffb23d6, the [test has failed](https://github.com/ironcore-dev/dpservice/actions/runs/8572357726/job/23494991604).

```
2024-04-05 15:54:32.996 2(grpc) I GRPC: Adding alias prefix, interface_id: vm2, prefix: 2a10:defe:dede:f4::1, prefix_len: 124 [../src/grpc/dp_async_grpc.cpp:327:FillRequest()]
176
2024-04-05 15:54:32.996 3(worker) I SERVICE: Value to be used, ptr: 0x7f2edaec2d90, ipv6: fc00:1::48:0:a8 [../src/dp_vnf.c:68:dp_add_vnf_value()]
177
2024-04-05 15:54:32.996 3(worker) I SERVICE: Value sending to table, ptr: 0x1a0d6a280, ipv6: fc00:1::48:0:a8 [../src/dp_vnf.c:82:dp_add_vnf_value()]
178
2024-04-05 15:54:32.996 3(worker) I SERVICE: Value stored in table, ptr: 0x1a0d6a280, ipv6: fc00:1::48:0:a8 [../src/dp_vnf.c:97:dp_add_vnf_value()]
```
...
```
2024-04-05 15:54:33.048 2(grpc) I GRPC: Removing alias prefix, interface_id: vm2, prefix: 2a10:defe:dede:f4::1, prefix_len: 124 [../src/grpc/dp_async_grpc.cpp:350:FillRequest()]
197
2024-04-05 15:54:33.048 3(worker) I SERVICE: Value found in table, ptr: 0x1a0d6a280, ipv6: fb00:1::48:0:a8 [../src/dp_vnf.c:243:dp_del_vnf_by_value()]
```
Given the fact that "Value stored" is actually retrieved using `rte_hash_lookup_data()` the same way as "Value found", I think something must corrupt the memory pointer holding the address in between these calls. 

But since this seems highly irregular, please find a logical error in my assumptions.

**Comment by PlagueCZ** [comment]
Thing just got even strange, I had a test fail with a different type of problem happen. Now it seems that `addvip` followed by immediage `getvip` returns corrupted data.
This actually happened in two separate occasions in different places of test (though both use the same copy-pasted code):
[test1](https://github.com/ironcore-dev/dpservice/actions/runs/8572784776/job/23497008512) [test2](https://github.com/ironcore-dev/dpservice/actions/runs/8554915274/job/23498782202)

Interesting part of the output (the correct/expected values are on the right side):
```
Differing items:
E      {'underlay_route': 'fc00:1::100:ac:0:f4'} != {'underlay_route': 'fc00:1::ac:0:f4'}
E      {'vip_ip': '172.20.0.0'} != {'vip_ip': '172.20.0.1'}
```

This seems to me that somehow the `.1` byte from VIP got written into the underlay address at byte 8 thus giving us the `:100` which is bytes 8 and 9, unless I'm mistaken. This is pretty similar to the problem previously where only a few bits were flipped.

So again we have a situation where the hash table returns the right pointer, but the data it is pointing to is somehow changed. What's more there is no other operation running in between the addition and lookup. And by previous tests I already confirmed that the data is correct after writing to a table and immediate lookup.

**Comment by PlagueCZ** [comment]
After an evening of unsucessful attempts to make pre-DPDK23 commit fail (14 tries), I tried a post-DPDK23 commit and got a failure in only two tries. Then trying the pre-DPDK23 again and nothing for another 10 tries. 

I am statistically leaning towards the possibility that the library is somehow the problem, although I have no idea how that would be possible.

---
There was one inscance of the pre-DPDK23 failing, but on a [packet contents](https://github.com/ironcore-dev/dpservice/actions/runs/8576284282/job/23509299170), so I am not fully convinced that it is related to our problems as some IPv6 packets are floating around during test and sometimes can break NeighborDiscovery

**Comment by PlagueCZ** [comment]
Another interesting find: I was unable to fail any run over the weekend nights. Yet my first attempt on Monday around noon resulted in a failure. 

I am unsure how the server running the test is used, but seems to indicate something load related?

**Comment by PlagueCZ** [comment]
Still happens, though now it's more rare.
https://github.com/ironcore-dev/dpservice/actions/runs/8721648897/attempts/1

**Comment by PlagueCZ** [comment]
The runner also failed now to download the docker image, I had to re-run for it to download it properly.

Then it downloaded but gRPC timed out, so this indicates an overloaded runner. 

May not be connected, just putting all the info here.

**Comment by guvenc** [comment]
For documenting the pre-allocated memory pattern (with hex values) used in test when the tests were failing with this problem described here.

MALLOC_PERTURB_=156: 0x9C
MALLOC_PERTURB_=87: 0x57
MALLOC_PERTURB_=22: 0x16
MALLOC_PERTURB_=248: 0xF8
MALLOC_PERTURB_=218: 0xDA
MALLOC_PERTURB_=78: 0x4E
MALLOC_PERTURB_=12: 0x0C
MALLOC_PERTURB_=178: 0xB2
MALLOC_PERTURB_=76: 0x4C

There seems to be no pattern between the values.

**Comment by PlagueCZ** [comment]
I also closed PR #515  that exists to run a test on a commit just before updating DPDK to 23, whch may be useful in the future.
